File: engine/sample_player.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <cstdlib>

using namespace std;

int team, number;

class Vector{
public:
    float x, y;
    Vector (float x = 0, float y = 0):x(x), y(y){}

    float dist(Vector other)
    {
        return sqrt((x - other.x) * (x - other.x) + (y - other.y) * (y - other.y));
    }

    float len()
    {
        return sqrt(x * x + y * y);
    }

    float angle()
    {
        float res;
        if (x == 0)
        {
            res = 3.1415/2;
            if (x < 0)
                res += 3.1415;
        }
        else
        {
            res = atan(1.0*y/x);
        }
        if (x < 0)
            res += 3.1415;
        return res;
    }
    Vector operator-(Vector other)
    {
        return Vector(x-other.x, y-other.y);
    }
    Vector operator+(Vector other)
    {
        return Vector(x+other.x, y+other.y);
    }
    Vector operator*(float num)
    {
        return Vector(x*num, y*num);
    }
};

vector<Vector> gplayers;
Vector last_cycle_dest(0, 0);

vector<Vector> get_players()
{
    vector <Vector> res;
    for (int i = 0; i < 10; i++)
    {
        float a, b;
        cin >> a >> b;
        if (team == 0)
        {
            a *= -1;
            b *= -1;
        }
        res.push_back(Vector(a, b));
    }
    return res;
}

Vector get_ballpos()
{
	Vector ballpos;
	cin >> ballpos.x >> ballpos.y;
	if (team == 0)
	{
		ballpos.x *= -1;
		ballpos.y *= -1;
	}
	return ballpos;
}

Vector get_target_position(int number, Vector ballpos)
{
    Vector pos(0, 0);
    if (number == 0)
    {
        pos = Vector(0, -40);
    }
    else if (number == 1)
    {
        pos = Vector(-20, -33);
    }
    else if (number == 2)
    {
        pos = Vector(20, -33);
    }
    else if (number == 3)
    {
        pos = Vector(-10, -10);
    }
    else if (number == 4)
    {
        pos = Vector(10, 0);
    }
//    if (ballpos.y > 0)
//        pos.y += ballpos.y * 3 / 4;

    return pos;
}

int get_ball_receiver(vector<Vector> players, Vector ballpos)
{
	vector<Vector> positions;
	float min_dist = 10e4;
	int receiver = 0;
	for (int i = 0; i <= 4; i++)
	{
		Vector pos = get_target_position(i, ballpos);
		if (pos.dist(ballpos) < min_dist)
		{
			min_dist = pos.dist(ballpos);
			receiver = i;
		}
	}
	return receiver;
}


float get_closest_op_dist(){
    //int other = 1-team;
    //cerr<<'heh'<<endl;
    float cmindist = 10e4;
    for(int i = 0; i <10; i++){
        cmindist = min(cmindist, (gplayers[team*5+number]-gplayers[i]).len());
    }
    return cmindist;
}

void move(Vector pos)
{
    if (team == 0)
    {
        pos.x *= -1;
        pos.y *= -1;
    }
    if((pos-last_cycle_dest).len() < 0.3 && get_closest_op_dist() < 3){
        pos = pos+Vector((rand()%2)*(-rand()%2), (rand()%2)*(-rand()%2));
        cerr<<"stuck"<<" "<<team<<" "<<number<<endl;    
    }
    last_cycle_dest = pos;
    cout << "move " << pos.x << " " << pos.y << endl;
    cout.flush();
}

void kick(Vector target, float power)
{
    if (team == 0)
    {
        target.x *= -1;
        target.y *= -1;
    }
    cout << "kick " << target.x << " " << target.y << " " << power << endl;
    cout.flush();
}

void kick_action()
{
    int a = rand() % 40 - 20;
    kick(Vector(10, 60), 1);
}


int main()
{
    srand(time(NULL));
    int game_state;
    cin >> team >> number;
    while (true)
    {
        vector <Vector> players = get_players();
        gplayers.clear();
        gplayers = players;
        //cerr<<get_closest_op_dist(players, team, number)<<endl;
        Vector selfpos = players[5*team+number];
        Vector ballpos = get_ballpos();
        cin >> game_state;

        int ball_reveiver = get_ball_receiver(players, ballpos);
        if (number == ball_reveiver)
        {
            if (selfpos.dist(ballpos) < 2)
            {
                kick_action();
            }
            else
                move(ballpos);
        }
        else
        {
            Vector position = get_target_position(number, ballpos);
            if (selfpos.dist(position) > 2)
            	move(position);
        }
    }
    return 0;
}

File: engine/sample_player.py
#!/usr/bin/python
from geometry import Vector, epsilon
from sys import stderr, stdout, exit
from utils import *
import time
from random import random


class Player():
    def run(self):
        self.last_move_pos = Vector(0, 0)
        self.team, self.number = map(int, raw_input().split(' '))
        while True:
            players = get_players(self.team)
            selfpos = players[5 * self.team + self.number]
            self.self_pos = selfpos
            self._players = players
            ballpos = get_ballpos(self.team)
            game_state = int(raw_input())
            # print >>stderr, 'GOT DATA AT %f'%(time.time())

            ball_receiver = self.get_ball_receiver(players, ballpos)
            if self.number == ball_receiver:
                if (selfpos - ballpos).len() < 2:
                    self.kick_action(ballpos)
                else:
                    self.move(ballpos)
            else:
                position = self.get_target_position(self.number, ballpos)
                if (selfpos - position).len() > 2:
                    self.move(position)

            # print >>stderr, 'SENT COMMAND AT %f'%(time.time())

    def get_target_position(self, number, ballpos):
        pos = Vector(0, 0)
        if number == 0:
            pos = Vector(0, -45)
        elif number == 1:
            pos = Vector(5, -30)
        elif number == 2:
            pos = Vector(-20, -20)
        elif number == 3:
            pos = Vector(20, -20)
        elif number == 4:
            pos = Vector(0, 0)
        return pos

    def move(self, position):
        if self.team == 0:
            position.x *= -1
            position.y *= -1
        if (self.last_move_pos-position).len() < 0.3 and self.nearest_opponent_dist() < 2:
            vec = Vector(-2 + int(random()*2) * 2, -2 + int(random() * 2) * 2)
            position = self.self_pos + vec
        print 'move %f %f' % (position.x, position.y)
        stdout.flush()
        self.last_move_pos = position

    def kick(self, target, power):
        if self.team == 0:
            target.x *= -1
            target.y *= -1
        print 'kick %f %f %f' % (target.x, target.y, power)
        stdout.flush()

    def nearest_opponent_dist(self):
        opp = 1 - self.team
        res = 10000
        for i in range(0, 5):
            print >>stderr, 'stuck'
            p = self._players[5*opp+i]
            if (p-self.self_pos).len() < res:
                res = (p-self.self_pos).len()
        return res

    def get_ball_receiver(self, players, ballpos):
        min_dist = 10000
        receiver = 0
        for i in range(5):
            pos = self.get_target_position(i, ballpos)
            if (pos - ballpos).len() < min_dist:
                min_dist = (pos - ballpos).len()
                receiver = i
        return receiver

    def kick_action(self, ballpos):
        if int(random() * 2) == 1:
            self.kick(Vector(10, 50), 1)
        else:
            self.kick(ballpos + Vector(10, 10), 1)


if __name__ == '__main__':
    p = Player()
    p.run()
